matrix-org/matrix-react-sdk
# Widget OpenID reauth implementation

Fixes https://github.com/vector-im/riot-web/issues/7153

The proposal for this is part of https://github.com/matrix-org/matrix-doc/issues/1286. The relevant snippet (and therefore draft) is available here: https://gist.github.com/turt2live/669bbffe882c8541f057daeb29e18b86

As a proof of concept, I've also created a Dimension widget which does nothing more than report your user ID back to you - all without using the `scalar_token` provided in the query string, and without you having to configure your client to point at a Dimension instance.

Testing procedure:
1. Create a new room to play with widgets in, or use an existing one
2. Using `/devtools`, create an `im.vector.modular.widgets` state event with state key `reauthwidget` and content of:
   ```json
   {
       "type": "customwidget",
	   "url": "https://dimension.t2bot.io/widgets/reauth",
	   "name": "OpenID Reauth Demo",
	   "data": {},
	   "id": "reauthwidget",
       "waitForIframeLoad": true,
	   "creatorUserId": "@travis:dev.t2host.io"
   }
   ```
   You may have to change the `creatorUserId` to be your user ID.
3. Close devtools and open the widget
4. Click the blue "Start OpenID auth" button. This should result in a popup in Riot.
5. Within the popup, decide if you want to always allow the widget and click either 'Allow' or 'Deny'
6. If you selected allow, the widget should update to show you your user ID.

To repeat the process, refresh the page. There's currently an unrelated bug that prevents reloading the widget from being enough to unstick the state: the whole page needs to be refreshed. This will be addressed in a different PR (https://github.com/matrix-org/matrix-react-sdk/pull/2799).

Source for the widget: https://github.com/turt2live/matrix-dimension/blob/master/web/app/widget-wrappers/reauth-example/reauth-example.component.ts

![image](https://user-images.githubusercontent.com/1190097/54470542-3e14c000-476f-11e9-8bed-5e365d02ba55.png)

----

This is being PR'd as a community member:
```
Signed-off-by: Travis Ralston <travis@t2bot.io>
```
